File: EFM_2024.dart

abstract class Employee {
  int _matricule = 0;
  String _nom = "";
  String _prenom = "";
  DateTime _dateNaissance = DateTime.now();
  Employee(this._matricule, this._nom, this._prenom, this._dateNaissance);

  int get matricule => _matricule;
  set matricule(int value) => _matricule = value;

  String get nom => _nom;
  set nom(String value) => _nom = value;

  String get prenom => _prenom;
  set prenom(String value) => _prenom = value;

  DateTime get dateNaissance => _dateNaissance;
  set dateNaissance(DateTime value) => _dateNaissance = value;
  @override
  String toString() {
    return "Matricule : $matricule \n Nom : $nom \n Prenom : $prenom \n Date de Naissance : $dateNaissance   ";
  }

  double getSalaire();
}

class Ouvrier extends Employee {
  DateTime _dateEntrer = new DateTime.now();
  static final double SMIG = 2500;

  Ouvrier(int matricule, String nom, String prenom, DateTime dateNaissance,
      this._dateEntrer)
      : super(matricule, nom, prenom, dateNaissance);

  @override
  String toString() {
    return super.toString() + "\n Date Enter $dateEntrer";
  }

  DateTime get dateEntrer => _dateEntrer;
  set dateEntrer(DateTime value) => _dateEntrer = value;

  @override
  double getSalaire() {
    double salaire = (DateTime.now().year - _dateEntrer.year) * SMIG;
    return (salaire > SMIG * 2) ? SMIG * 2 : salaire;
  }
}

class Cadre extends Employee {
  int _indice = 0;
  Cadre(int matricule, String nom, String prenom, DateTime dateNaissance,this._indice) : super(matricule, nom, prenom, dateNaissance);
  int get indice => _indice;
  set indice(int value) => _indice = value;

  @override
  String toString() {
    return super.toString() + " \n Indice : $indice";
  }

  @override
  double getSalaire() {
    switch (_indice) {
      case 1:
        return 13000;
      case 2:
        return 15000;
      case 3:
        return 17000;
      case 4:
        return 20000;
      default:
        return 0;
    }
  }
}

class Patron extends Employee {
  static final int Chiffre_Affaire = 0;
  double _pourcentage = 0;
  Patron(int matricule, String nom, String prenom, DateTime dateNaissance,
      this._pourcentage) : super(matricule, nom, prenom, dateNaissance);
  double get pourcentage => _pourcentage;
  set pourcentage(double value) => _pourcentage = value;

  @override
  String toString() {
    return super.toString() + "\n Pourcentage : $_pourcentage";
  }

  @override
  double getSalaire() {
    return Chiffre_Affaire * pourcentage / 100;
  }
}

void main(List<String> args) {
  Cadre cadre = Cadre(1, "nom", "prenom", DateTime(2000, 10, 4), 1);
  print(cadre);
  print(cadre.getSalaire());

}

File: EFM_2024.kt
open abstract class Empolye(var matricule:Int , var nom:String , var prenom:String,var dateNaissance:Date){
    open override fun toString():String{
        return "matricule $matricule \n le nom : $nom \n le prenom $prenom \n date de naissance : $dateNaissance"
    }
    abstract fun getSalaire():Double;

}

class Ouvrier( matricule:Int ,  nom:String ,  prenom:String, dateNaissance:Date,var dateEntre:Date) : Empolye( matricule,  nom ,  prenom,dateNaissance){

    companion object {
         const val SMIG : Double=2500.0
    }
          override fun toString():String{
               return super.toString()+"\ndateEntre:$dateEntre";
          }
          
          override fun getSalaire():Double{
                 val a = LocalDate.now().year-dateEntre.year;
                 val salaire=SMIG*a*100 
                 return salaire if salaire<2*SMIG else 2*SMIG 

          }
}
class Cadr( matricule:Int ,  nom:String ,  prenom:String, dateNaissance:Date,var indice:Int) : Empolye( matricule,  nom ,  prenom,dateNaissance){
     override fun toString():String{
               return super.toString()+"\nindice:$indice";
          }
           override fun getSalaire():Double{
                 
                 val salaire = when(indice){
                    1 -> 13000
                    2 -> 15000
                    3 -> 17000
                    4 ->  20000
                    else -> 0
                 }

                 when(indice){
                    1 -> return 13000 
                    2 -> return 15000
                    3 -> return 17000
                    4 -> return  20000
                    else -> return 0
                 }

          }
}
class Patron(matricule:Int ,  nom:String ,  prenom:String, dateNaissance:Date,var pourcentage:Double):Empolye(matricule,  nom ,  prenom,dateNaissance){
     override fun toString():String{
               return super.toString()+"\nPourcentage: $pourcentage\nChiffre d affaire:$chiffreAffaire";
          }
          companion object{
            final var chiffreAffaire=0
          }
          override fun getSalaire(){
            return chiffreAffaire*pourcentage/100
          }
}

Empolye.calculerAge():int{
    return LocalDate.now().year - this.dateNaissance.year
}